### PR TITLE
Remove code duplication in initialization of temperature and composition

### PR DESCRIPTION
**Pull request by class4kayaker**
Addresses issue #1145. 

The unification on the standard of the n_compos+1 loop is due to wanting to reduce logic duplication.

This pull request also adds a test for the composition normalization due to no existing test using said feature.

### CODE REVIEW

File: source/simulator/initial_conditions.cc [hunk 1]
```diff
@@ -214,6 +91,55 @@ namespace aspect
 
         std::vector<types::global_dof_index> local_dof_indices (finite_element.dofs_per_cell);
 
+#if DEAL_II_VERSION_GTE(8,5,0)
```
Comment (gassmoeller) [inline source/simulator/initial_conditions.cc]: The tester currently only tests deal.II 8.4.0. Did you test that this changed code works with 8.5.0.pre? If you do not have a development version of deal.II I could also do that for you. Let me know.

File: source/simulator/initial_conditions.cc [hunk 2]
```diff
@@ -214,6 +91,55 @@ namespace aspect
...
+        const VectorFunctionFromScalarFunctionObject<dim, double> &advf_init_function =
+          (advf.is_temperature()?
```
Comment (gassmoeller) [inline source/simulator/initial_conditions.cc]: Please put the ? and : operators on separate lines.

File: source/simulator/initial_conditions.cc [hunk 3]
```diff
@@ -214,6 +91,55 @@ namespace aspect
...
+                                                               introspection.n_components):
+           VectorFunctionFromScalarFunctionObject<dim, double>(std_cxx11::bind(&CompositionalInitialConditions::Interface<dim>::initial_composition,
+                                                                               std::ref(*compositional_initial_conditions),
+                                                                               std_cxx11::_1,
+                                                                               n-1),
+                                                               introspection.component_indices.compositional_fields[n-1],
+                                                               introspection.n_components));
+
+        const ComponentMask advf_mask =
+          (advf.is_temperature()?
```
Comment (gassmoeller) [inline source/simulator/initial_conditions.cc]: Same here

File: source/simulator/initial_conditions.cc [hunk 4]
```diff
@@ -214,6 +91,55 @@ namespace aspect
...
+                fe_values.reinit (cell);
+
+                // Go through the support points for each dof
+                for (unsigned int i=0; i<finite_element.base_element(base_element).dofs_per_cell; ++i)
+                  {
+                    // if it is specified in the parameter file that the sum of all compositional fields
+                    // must not exceed one, this should be checked
+                    double sum = 0;
+                    for (unsigned int m=0; m<parameters.normalized_fields.size(); ++m)
+                      sum += compositional_initial_conditions->initial_composition(fe_values.quadrature_point(i),parameters.normalized_fields[m]);
```
Comment (gassmoeller) [inline source/simulator/initial_conditions.cc]: Please break this line after the first parameter of initial_composition (simpler reading).

File: source/simulator/initial_conditions.cc [hunk 5]
```diff
@@ -214,6 +91,55 @@ namespace aspect
...
+                    if (std::abs(sum) > 1.0+1e-6)
```
Comment (gassmoeller) [inline source/simulator/initial_conditions.cc]: I think we could replace 1e-6 by `std::numeric_limits<double>::epsilon()`.

Comment (gassmoeller) [inline source/simulator/initial_conditions.cc]: And there is another occurence in the other code path, please replace that as well. I realize it was there before your change, but this is the opportunity to clean up :smile:

File: source/simulator/initial_conditions.cc [hunk 6]
```diff
@@ -214,6 +91,55 @@ namespace aspect
...
+           introspection.component_masks.temperature:
+           introspection.component_masks.compositional_fields[n-1]);
+
+        VectorTools::interpolate(*mapping,
+                                 dof_handler,
+                                 advf_init_function,
+                                 initial_solution,
+                                 advf_mask);
+
+        if (n == 1 && parameters.normalized_fields.size()>0)
```
Comment (gassmoeller) [inline source/simulator/initial_conditions.cc]: In the other code path (for deal.II 8.4.0) the condition is written the other way `if (parameters.normalized_fields ... && n == 1)`. Not very important but for easier review: please order the conditions in this `if` the same way.

**Comment by gassmoeller** [comment]
Very nice, thanks for the cleaning! I have just a few style issues, and the question if this runs with deal.II 8.5.0.pre. Let me know if you need help with that. Other than that this looks good.

File: source/simulator/initial_conditions.cc (hunk 1 above)
Comment (class4kayaker) [inline source/simulator/initial_conditions.cc]: Yes, I did check that the changed code worked for 8.5.0-pre. I will check again with the most recent version of deal.II though.

**Comment by class4kayaker** [comment]
Tested with 8.5.0.pre, and there are some variations in behavior, but I believe that these are due to the version change having other effects. I have confirmed that sol_cx_2 and sol_cx_4 tests are passing with the update. Is there a good MMS composition test I can use?

**Comment by gassmoeller** [comment]
Not sure what you mean by MMS? Apart from that question this is ready to merge.

**Comment by class4kayaker** [comment]
I was asking if there was a good initial condition test for composition since sol_cx does not explicitly test the values when I checked.

**Comment by gassmoeller** [comment]
The composition_active test should be good. Does that work with the recent deal.II?

**Comment by class4kayaker** [comment]
The composition_active test passes with deal.II commit dealii/dealii@7b35fdd, which was the master as August 9, 2016. So that should be good. I wanted a good double check because the passive_comp test did not pass due to things that were reasonable local variations.

**Comment by gassmoeller** [comment]
Ok, then lets call it a day. It seems reasonable and passes all tests we tested. Thanks again!

**Pull request merged**